src/libraries/System.Private.CoreLib/src/System/Collections/Generic/ArraySortHelper.cs

```diff
@@ -473,8 +473,8 @@ private static int PickPivotAndPartition(Span<T> keys)
             {
                 if (pivot == null)
                 {
-                    while (Unsafe.IsAddressLessThan(ref leftRef, ref nextToLastRef) && (leftRef = ref Unsafe.Add(ref leftRef, 1)) == null) ;
-                    while (Unsafe.IsAddressGreaterThan(ref rightRef, ref zeroRef) && (rightRef = ref Unsafe.Add(ref rightRef, -1)) == null) ;
+                    while (Unsafe.IsAddressLessThan(ref leftRef!, ref nextToLastRef) && (leftRef = ref Unsafe.Add(ref leftRef, 1)) == null) ;
+                    while (Unsafe.IsAddressGreaterThan(ref rightRef!, ref zeroRef) && (rightRef = ref Unsafe.Add(ref rightRef, -1)) == null) ;
```

Comment:
I'm not understanding why the compiler flags this. Why does the compiler think these methods could assign null to the ref?

Reply:
I believe the warning is for the value of `ref` argument when the method is called. On the second time through the loop, `leftRef` is assumed to be `null` based on `&& (leftRef = ...) == null`.

Reply:
I see.  Is the right fix here then to actually annotate both arguments with [AllowNull]?

Reply:
> Is the right fix here then to actually annotate both arguments with [AllowNull]?

The methods in `Unsafe` simply take an unconstrained `T` which seems like it should be sufficient for all callers. Should the `leftRef`, etc. locals be declared as `ref T?` here instead?

Reply:
> Should the `leftRef`, etc. locals be declared as `ref T?` here instead?

Perhaps not. That would just move the warnings to the initializers of those locals.

Reply:
> Should the leftRef, etc. locals be declared as ref T? here instead?

Wouldn't that be a problem if a `T t;` were passed in, with the compiler then complaining that the implementation might assign null to it?

Reply:
Right, hence my question about [AllowNull].  Does that make sense?

Reply:
This is similar to the `Volatile.Read` APIs, where ideally we could change `ref T` to `in T`, but that would be a breaking API change.